cgo/cgo.go

```
@@ -35,37 +36,41 @@ func initialize(cLogDir *C.char) *C.char {
 	walletsMtx.Lock()
 	defer walletsMtx.Unlock()
 	if wallets != nil {
-		return cString("duplicate initialization")
+		return errCResponse("duplicate initialization")
 	}
 
 	logDir := goString(cLogDir)
 	logSpinner, err := assetlog.NewRotator(logDir, "dcrwallet.log")
 	if err != nil {
-		return cStringF("error initializing log rotator: %v", err)
+		return errCResponse("error initializing log rotator: %v", err)
 	}
 
 	logBackend = newParentLogger(logSpinner)
 	err = dcr.InitGlobalLogging(logDir, logBackend)
 	if err != nil {
-		return cStringF("error initializing logger for external pkgs: %v", err)
+		return errCResponse("error initializing logger for external pkgs: %v", err)
 	}
 
+	logMtx.Lock()
 	log = logBackend.SubLogger("[APP]")
 	log.SetLevel(slog.LevelTrace)
+	logMtx.Unlock()
 
 	ctx, cancelCtx = context.WithCancel(context.Background())
 	wallets = make(map[string]*wallet)
 
-	log.Info("libwallet cgo initialized")
-	return nil
+	return successCResponse("libwallet cgo initialized")
 }
 
 //export shutdown
-func shutdown() {
+func shutdown() *C.char {
+	logMtx.RLock()
+	log.Debug("libwallet cgo shutting down")
+	logMtx.RUnlock()
 	walletsMtx.Lock()
 	for _, wallet := range wallets {
 		if err := wallet.CloseWallet(); err != nil {
-			wallet.log.Errorf("close wallet error: %v", err)
+			errCResponse("close wallet error: %v", err)
```

Do you mean to return this here?

---

Oh wow, idk. what do we do here. We want to go through all of them. Do we want to return an error if one errors or just log? I don't think cake can support multiple wallets so maybe not a problem if we return err on any error...

---

Can it support multiple?

---

Yes, it can, I believe.